File: app/num_words.py
"""Convert monetary amounts to bilingual (Polish/English) word form."""

from decimal import Decimal

from num2words import num2words


def _pl_form(n: int, forms: tuple[str, str, str]) -> str:
    """Return correct Polish noun form based on number.

    forms: (form_1, form_2_4, form_5plus)
    Examples:
        1 złoty, 2 złote, 5 złotych
        1 euro,  2 euro,  5 euro
    """
    form1, form2_4, form5 = forms
    if n == 1:
        return form1
    last_two = n % 100
    last_one = n % 10
    if 12 <= last_two <= 14:
        return form5
    if 2 <= last_one <= 4:
        return form2_4
    return form5


# Currency definitions: (pl_main_forms, pl_subunit, en_main, en_subunit)
_CURRENCIES: dict[str, tuple[tuple[str, str, str], str, str, str]] = {
    "EUR": (("euro", "euro", "euro"), "centów", "euro", "cent"),
    "PLN": (("złoty", "złote", "złotych"), "groszy", "złoty", "grosz"),
    "USD": (("dolar", "dolary", "dolarów"), "centów", "dollar", "cent"),
    "GBP": (("funt", "funty", "funtów"), "pensów", "pound", "penny"),
}


def amount_in_words(amount_str: str, currency: str) -> tuple[str, str]:
    """Return (polish, english) word representation of a monetary amount.

    Polish format: "{words} {currency_noun} {cents}/100"
    English format: "{words} {currency} and {cent_words} {subunit}(s)"
    """
    value = Decimal(amount_str)
    whole = int(value)
    cents = int(round((value - whole) * 100))

    pl_words = num2words(whole, lang="pl")
    en_words = num2words(whole, lang="en")

    if currency in _CURRENCIES:
        pl_forms, pl_subunit, en_main, en_subunit = _CURRENCIES[currency]
        pl_currency = _pl_form(whole, pl_forms)
        pl_result = f"{pl_words} {pl_currency} {cents}/100"

        en_subunit_plural = en_subunit + "s" if cents != 1 else en_subunit
        en_cent_words = num2words(cents, lang="en")
        en_result = f"{en_words} {en_main} and {en_cent_words} {en_subunit_plural}"
    else:
        # Unknown currency: use code as-is
        pl_result = f"{pl_words} {currency} {cents}/100"
        en_cent_words = num2words(cents, lang="en")
        en_result = f"{en_words} {currency} and {en_cent_words} cents"

    return pl_result, en_result

File: web2/src/num-words.ts
// Convert monetary amounts to bilingual (Polish/English) word form.
// Port of app/num_words.py — uses n2words instead of Python's num2words.
//
// Drobna różnica formatu względem Pythonowego num2words:
//   Python (en) wstawia przecinek po "thousand": "one thousand, two hundred…"
//   n2words (en-GB) tylko spację: "one thousand two hundred and …"
// Treściowo identyczne, więc akceptujemy.

import { toCardinal as toCardinalPL } from "n2words/pl-PL";
import { toCardinal as toCardinalEN } from "n2words/en-GB";

/** Polish noun forms: pick correct form based on number. */
function plForm(n: number, forms: [string, string, string]): string {
  const [form1, form2_4, form5] = forms;
  if (n === 1) return form1;
  const lastTwo = n % 100;
  const lastOne = n % 10;
  if (lastTwo >= 12 && lastTwo <= 14) return form5;
  if (lastOne >= 2 && lastOne <= 4) return form2_4;
  return form5;
}

interface CurrencySpec {
  pl: [string, string, string];
  plSubunit: string;
  en: string;
  enSubunit: string;
}

const CURRENCIES: Record<string, CurrencySpec> = {
  EUR: { pl: ["euro", "euro", "euro"], plSubunit: "centów", en: "euro", enSubunit: "cent" },
  PLN: { pl: ["złoty", "złote", "złotych"], plSubunit: "groszy", en: "złoty", enSubunit: "grosz" },
  USD: { pl: ["dolar", "dolary", "dolarów"], plSubunit: "centów", en: "dollar", enSubunit: "cent" },
  GBP: { pl: ["funt", "funty", "funtów"], plSubunit: "pensów", en: "pound", enSubunit: "penny" },
};

/** Returns [polish, english] word representation of a monetary amount. */
export function amountInWords(amountStr: string, currency: string): [string, string] {
  const value = Number(amountStr);
  if (!Number.isFinite(value)) return ["", ""];
  const whole = Math.trunc(value);
  const cents = Math.round(Math.abs(value - whole) * 100);

  const plWords = toCardinalPL(whole);
  const enWords = toCardinalEN(whole);

  const spec = CURRENCIES[currency.toUpperCase()];
  if (spec) {
    const plCurrency = plForm(whole, spec.pl);
    const plResult = `${plWords} ${plCurrency} ${cents}/100`;
    const enSubunitPlural = cents === 1 ? spec.enSubunit : `${spec.enSubunit}s`;
    const enCentWords = toCardinalEN(cents);
    const enResult = `${enWords} ${spec.en} and ${enCentWords} ${enSubunitPlural}`;
    return [plResult, enResult];
  }

  // Unknown currency: use ISO code as-is
  const enCentWords = toCardinalEN(cents);
  return [
    `${plWords} ${currency} ${cents}/100`,
    `${enWords} ${currency} and ${enCentWords} cents`,
  ];
}
